packages/ses/src/commons.js

```diff
@@ -174,6 +176,89 @@ export const arraySome = uncurryThis(arrayPrototype.some);
 export const arraySort = uncurryThis(arrayPrototype.sort);
 export const iterateArray = uncurryThis(arrayPrototype[iteratorSymbol]);
 //
+export const arrayBufferSlice = uncurryThis(arrayBufferPrototype.slice);
+const { asUintN: bigIntAsUintN } = BigInt;
+const { structuredClone } = globalThis;
+export const arrayBufferTransferToFixedLength =
+  // @ts-expect-error absent from Node <= 20 and not yet in TS type
+  // eslint-disable-next-line no-nested-ternary
+  arrayBufferPrototype.transferToFixedLength
+    ? // @ts-expect-error absent from Node <= 20 and not yet in TS type
+      uncurryThis(arrayBufferPrototype.transferToFixedLength)
+    : structuredClone
+      ? /**
+         * @param {ArrayBuffer} arrayBuffer
+         * @param {number} newLength
+         * @returns {ArrayBuffer}
+         */
+        (arrayBuffer, newLength = arrayBuffer.byteLength) => {
+          // There is no `transferToFixedLength` on Node <= 20, but there
+          // is web-standard `structuredClone` on Node >= 17, on all modern
+          // browsers, and on many other JS platforms.
+          // In those cases, we first use `structuredClone` to get a fresh
+          // buffer with exclusive access to the underlying data, while
+          // detaching it from the original `arrayBuffer`.
+
+          newLength = +newLength;
+          bigIntAsUintN(newLength, 0n);
+
+          const newBuffer = /** @type {ArrayBuffer} */ (
+            structuredClone(arrayBuffer, {
+              transfer: [arrayBuffer],
+            })
+          );
+          if (newLength >= newBuffer.byteLength) {
+            return newBuffer;
+          }
+          // If the requested length is shorter than the length of `buffer`,
+          // we use `slice` to shorted the returned result, but at the cost
+          // of an extra copy.
+          //
+          // `slice` accepts negative arguments but `transferToFixedLength`
+          // does not...
+          // get at the underlying ToIndex operation through `BigInt.asUintN`
+          // (avoiding the redundant allocation of e.g. `ArrayBuffer(newLength)`)
+          // and ToNumber through unary `+` (rather than `Number(newLength)`,
+          // which fails to reject BigInts).
+          //
+          // On platforms like Node 20
+          // - without`tranferToFixedLength` or `transfer`
+          // - with `structuredClone`
+          // - with `resize`
+          //
+          // It might seem like we could avoid the extra copy by
+          // `newBuffer.resize(newLength)`. But `structuredClone`
+          // makes ArrayBuffers that are not resizable.
+          return arrayBufferSlice(newBuffer, 0, newLength);
+        }
```

Comment:
```suggestion
        (arrayBuffer, newLength = arrayBufferByteLength(arrayBuffer)) => {
          // There is no `transferToFixedLength` on Node <= 20, but there
          // is web-standard `structuredClone` on Node >= 17, on all modern
          // browsers, and on many other JS platforms.
          // In those cases, we first use `structuredClone` to get a fresh
          // buffer with exclusive access to the underlying data, while
          // detaching it from the original `arrayBuffer`.

          // Before looking at actual arguments, verify that the input is an
          // ArrayBuffer.
          arrayBufferByteLength(arrayBuffer);

          // Calculate ToIndex(newLengthAsNumber) using `BigInt.asUintN`,
          // first getting newLengthAsNumber as ToNumber(newLength) using unary `+`
          // (rather than `Number(newLength)`, which fails to reject BigInts).
          newLength = +newLength;
          bigIntAsUintN(newLength, 0n);

          const newBuffer = /** @type {ArrayBuffer} */ (
            structuredClone(arrayBuffer, {
              transfer: [arrayBuffer],
            })
          );

          // We might already have what we need.
          // NOTE: The check for resizability is necessary to correctly emulate
          // ArrayBuffer.prototype.transferToFixedLength, but a more narrow
          // function specialized to e.g. transferToImmutable could skip it.
          if (newLength === arrayBufferByteLength(newBuffer) && !arrayBufferResizable(newBuffer)) {
            return newBuffer;
          }

          // We might be able to copy some or all of the contents.
          if (newLength <= arrayBufferByteLength(newBuffer)) {
            const copied = arrayBufferSlice(newBuffer, 0, newLength);
            const copiedLength = arrayBufferByteLength(copied);
            if (copiedLength !== newLength) {
              throw RangeError(`internal: length ${copiedLength} should have been ${newLength}`);
            }
            return copied;
          }

          // We need a bigger boat.
          // NOTE: Uint8Array must be extracted from globalThis.
          const view = new Uint8Array(newLength);
          typedArraySet(view, new Uint8Array(newBuffer));
          return typedArrayBuffer(view);
```